swap/src/api/request.rs

```diff
@@ -0,0 +1,679 @@
+use crate::api::Context;
+use crate::bitcoin::{Amount, TxLock};
+use crate::cli::{list_sellers, EventLoop, SellerStatus};
+use crate::libp2p_ext::MultiAddrExt;
+use crate::network::quote::{BidQuote, ZeroQuoteReceived};
+use crate::network::swarm;
+use crate::protocol::bob;
+use crate::protocol::bob::{BobState, Swap};
+use crate::{bitcoin, cli, monero, rpc};
+use anyhow::{bail, Context as AnyContext, Result};
+use libp2p::core::Multiaddr;
+use qrcode::render::unicode;
+use qrcode::QrCode;
+use serde_json::json;
+use std::cmp::min;
+use std::convert::TryInto;
+use std::future::Future;
+use std::sync::Arc;
+use std::time::Duration;
+use std::net::SocketAddr;
+use tokio::sync::broadcast;
+use uuid::Uuid;
+
+#[derive(PartialEq, Debug)]
+pub struct Request {
+    pub params: Params,
+    pub cmd: Method,
+    pub shutdown: Shutdown,
+}
+
+impl Shutdown {
+    pub fn new(notify: broadcast::Receiver<()>) -> Shutdown {
+        Shutdown {
+            shutdown: false,
+            notify,
+        }
+    }
+
+    /// Returns `true` if the shutdown signal has been received.
+    pub fn is_shutdown(&self) -> bool {
+        self.shutdown
+    }
+
+    /// Receive the shutdown notice, waiting if necessary.
+    pub async fn recv(&mut self) {
+        // If the shutdown signal has already been received, then return
+        // immediately.
+        if self.shutdown {
+            return;
+        }
+
+        // Cannot receive a "lag error" as only one value is ever sent.
+        let _ = self.notify.recv().await;
+
+        self.shutdown = true;
+
+        // Remember that the signal has been received.
+    }
+}
+
+#[derive(Debug)]
+pub struct Shutdown {
+    shutdown: bool,
+    notify: broadcast::Receiver<()>,
+}
+
+impl PartialEq for Shutdown {
+    fn eq(&self, other: &Shutdown) -> bool {
+        self.shutdown == other.shutdown
+    }
+}
+
+#[derive(Default, PartialEq, Debug)]
+pub struct Params {
+    pub seller: Option<Multiaddr>,
+    pub bitcoin_change_address: Option<bitcoin::Address>,
+    pub monero_receive_address: Option<monero::Address>,
+    pub rendezvous_point: Option<Multiaddr>,
+    pub swap_id: Option<Uuid>,
+    pub amount: Option<Amount>,
+    pub server_address: Option<SocketAddr>,
+    pub address: Option<bitcoin::Address>,
+}
```

Comment:
Is there a better way to structure this?
We are making all fields optional here which requires us to later again .unwrap() to make sure they exist. Could we maybe combine the Method and Params struct to assign all the required fields to each Method individually?

@delta1 Perhaps you have some input

Reply:
I'll play around with it in the next days to see if it can be structured in a cleaner way